lib/Sema/ConstraintSystem.cpp

```diff
@@ -1175,7 +1175,8 @@ ConstraintSystem::getTypeOfMemberReference(
          func->hasArchetypeSelf())) {
       openedType = openedType->replaceCovariantResultType(
                      baseObjTy,
-                     func->getNumParameterLists());
+                     func->getNumParameterLists(),
+                     !baseObjTy->getAnyOptionalObjectType());
```

Comment:
Instead of changing the `preserveOptionality` argument, it seems like we should look through optional types on the `baseObjTy` and pass down the result... but be careful when we're actually calling an initializer on `Optional` or `ImplicitlyUnwrappedOptional` itself!